src/libraries/Common/src/System/Security/Cryptography/MLDsa.cs

```diff
@@ -2218,5 +2399,78 @@ internal static void ThrowIfNotSupported()
         }
 
         private delegate TResult ExportPkcs8PrivateKeyFunc<TResult>(ReadOnlySpan<byte> pkcs8);
+
+        private sealed class DefaultMLDsaMuHash : MLDsaMuHash
+        {
+            private byte[] _keyState;
+            private Shake256 _shake;
+
+            private DefaultMLDsaMuHash(DefaultMLDsaMuHash toClone)
+                : base(toClone.Key)
+            {
+                _shake = toClone._shake.Clone();
+
+                // The key state is never updated after the initial computation,
+                // so it doesn't need a full clone.
+                _keyState = toClone._keyState;
+            }
+
+            internal DefaultMLDsaMuHash(MLDsa mldsa, ReadOnlySpan<byte> context)
+                : base(mldsa)
+            {
+                // FIPS 204, 5.3, Algorithm 3:
+                // M' = BytesToBits(IntegerToBytes(0, 1) || IntegerToBytes(|ctx|, 1) || ctx) || M
+                // FIPS 204, 6.3, Algorithm 8:
+                // tr = H(pk, 64)
+                // mu = H(BytesToBits(tr) || M')
+                //
+                // So on Reset we need to know tr (64 bytes), the "pure" zero, the length of the context, and the context.
+                _keyState = new byte[64 + 1 + 1 + context.Length];
+                _shake = new Shake256();
+
+                Span<byte> tr = _keyState.AsSpan(0, 64);
+
+                using (CryptoPoolLease pk = CryptoPoolLease.Rent(mldsa.Algorithm.PublicKeySizeInBytes, skipClear: true))
+                {
+                    mldsa.ExportMLDsaPublicKey(pk.Span);
+                    _shake.AppendData(pk.Span);
+                    _shake.GetHashAndReset(tr);
+                }
+
+                // The "pure" signature marker.
+                // It's already zero (because we used new[]), but let's set it anyways.
+                _keyState[64] = 0;
+                _keyState[65] = checked((byte)context.Length);
+                context.CopyTo(_keyState.AsSpan(66));
+                _shake.AppendData(_keyState);
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                if (disposing)
+                {
+                    _shake.Dispose();
+                    _keyState = null!;
+                    _shake = null!;
+                }
+            }
+
+            protected override void AppendDataCore(ReadOnlySpan<byte> data) =>
+                _shake.AppendData(data);
+
+            protected override void GetCurrentHashCore(Span<byte> destination) =>
+                _shake.GetCurrentHash(destination);
+
+            protected override void GetHashAndResetCore(Span<byte> destination)
+            {
+                _shake.GetHashAndReset(destination);
+                _shake.AppendData(_keyState);
```

Comment:
Up to you if you think it is worth it, but this seems like an ideal place where `Clone` can be a checkpoint. It would mean having two Shake256 instances though, but you could drop the `_keyState` array.

You have a source `Shake256` that appends `_keyState`. Call it `_initial` or something. The constructor does `_shake = _initial.Clone()`.

`GetHashAndResetCore()` does the `GetHashAndReset`, disposes of it, and clones the _initial again.

I am not sure this is better than just keeping the state around. It does mean reset doesn't have to re-append every time.

Reply:
Yeah, if OSSL exposes the hash accumulator in the future, that's how I expected to do it.  .NET 8 doesn't have Clone, so it's hard to do in the base class... at least in the compat package.